### PR TITLE
move tailsitter-specific code to new AttitudeControl_TS subclass

### PR DESCRIPTION
**Pull request by kd0aij**
replaces https://github.com/ArduPilot/ardupilot/pull/15195

Adds an overloaded version of relax_attitude_controllers with one boolean argument which simply calls the original no-arg version in the base class.  The new subclass inherits from AC_AttitudeControl_Multi and overrides the one-argument version of  relax_attitude_controllers to implement special behavior for tilt-vectored tailsitters.

The implementation of the existing tailsitter specific method input_euler_rate_yaw_euler_angle_pitch_bf_roll is also moved from the base class to the subclass.

This is a fix for the problem referenced in ArduPilot/ardupilot_wiki#2990
It allows the motors to tilt up in all vtol modes (not just qstabilize) prior to raising the throttle, and prevents prop strikes on arming.

### CODE REVIEW

**Comment by kd0aij** [comment]
@tridge This is the alternative approach to tailsitter attitude control  specialization preferred by both Randy and Leonard

**Comment by kd0aij** [comment]
this fix needs to be rebased on https://github.com/ArduPilot/ardupilot/pull/15301 once that goes in

**Comment by kd0aij** [comment]
@Hwurzburg I just retested this in sitl with the SkyCat2. Seems to behave the same as the original branch you flight tested.

**Comment by kd0aij** [comment]
@IamPete1 It looks like this branch always had the fix to hold_stabilize, unless I"m confused.

**Comment by kd0aij** [comment]
rebased
build/Pixracer/bin/arduplane.apj:    "image_size": 1429312,
master: "image_size": 1428688, delta 624 bytes

File: ArduPlane/quadplane.cpp [hunk 1]
```diff
@@ -708,7 +709,11 @@ bool QuadPlane::setup(void)
         AP_BoardConfig::config_error("Unable to allocate %s", "ahrs_view");
     }
 
-    attitude_control = new AC_AttitudeControl_Multi(*ahrs_view, aparm, *motors, loop_delta_t);
+    if (motor_class == AP_Motors::MOTOR_FRAME_TAILSITTER) {
```
Comment (kd0aij) [inline ArduPlane/quadplane.cpp:712]: does this need to include copter tailsitters also?

Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: yes, fixed

**Comment by kd0aij** [comment]
@robustini I think this is the version of https://github.com/ArduPilot/ardupilot/pull/15195 which may eventually go into master. I'd appreciate it if you could test to see whether there is any breakage for non-vectored tailsitters, and what issues remain to be solved.
I'll look into negating any ill effects from Q_TRIM_PITCH on takeoff.

**Comment by robustini** [comment]
> 
> 
> @robustini I think this is the version of #15195 which may eventually go into master. I'd appreciate it if you could test to see whether there is any breakage for non-vectored tailsitters, and what issues remain to be solved.
> I'll look into negating any ill effects from Q_TRIM_PITCH on takeoff.

No problem found on non-vectored (SITL and Realflight).

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -741,6 +748,11 @@ bool QuadPlane::setup(void)
     pos_control->set_dt(loop_delta_t);
     attitude_control->parameter_sanity_check();
 
+    // TODO: update this if servo function assignments change
+    // used by relax_attitude_control() to control special behavior for vectored tailsitters
+    _is_vectored = ((motor_class == AP_Motors::MOTOR_FRAME_TAILSITTER) &&
+           (!is_zero(tailsitter.vectored_hover_gain) && SRV_Channels::function_assigned(SRV_Channel::k_tiltMotorLeft)));
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:754]: since were only doing this one time you might as well also check tiltMotorRight

**Review by IamPete1 — APPROVED** [review]

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:754]: but you were the one who made me take that check out :)

Comment (IamPete1) [reply ArduPlane/quadplane.cpp:754]: Maybe when it was once a loop? I meant check for one or the other. 
```suggestion
           (!is_zero(tailsitter.vectored_hover_gain) && (SRV_Channels::function_assigned(SRV_Channel::k_tiltMotorLeft) || SRV_Channels::function_assigned(SRV_Channel::k_tiltMotorRIght)));
```

Comment (kd0aij) [reply ArduPlane/quadplane.cpp:754]: I hope I never had that in a loop, but maybe...
What happens if you "commit suggestion". Does it do a weird merge commit? I pushed the "resolve inconsistencies" button recently, and didn't like the result at all.

Comment (IamPete1) [reply ArduPlane/quadplane.cpp:754]: You just get a extra commit on top, the message is abit random usually so I tend to then pull it down squash and push back up. 

Maybe it was on some slower check or something, to be honest I have no idea, maybe I'm just fickle ;)

Comment (kd0aij) [reply ArduPlane/quadplane.cpp:754]: Thanks, I'll give that a try then.

File: libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+/// @brief   ArduCopter attitude control library
+
+#include "AC_AttitudeControl_Multi.h"
+
+
+class AC_AttitudeControl_TS : public AC_AttitudeControl_Multi {
+public:
+	AC_AttitudeControl_TS(AP_AHRS_View &ahrs, const AP_Vehicle::MultiCopter &aparm, AP_MotorsMulticopter& motors, float dt);
+
+	// empty destructor to suppress compiler warning
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:13]: little white space issue

File: libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	virtual ~AC_AttitudeControl_TS() {}
+
+    // Ensure attitude controller have zero errors to relax rate controller output
+    // Relax only the roll and yaw rate controllers if exclude_pitch is true
+    virtual void relax_attitude_controllers(bool exclude_pitch) override;
+    virtual void input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool plane_controls, float body_roll_cd, float euler_pitch_cd, float euler_yaw_rate_cds) override;
+    void set_is_vectored(bool v) { _is_vectored = v; }
+    bool get_is_vectored(void) { return _is_vectored; }
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:21]: get_is_vectored() could be const

File: libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	AC_AttitudeControl_TS(AP_AHRS_View &ahrs, const AP_Vehicle::MultiCopter &aparm, AP_MotorsMulticopter& motors, float dt);
+
+	// empty destructor to suppress compiler warning
+	virtual ~AC_AttitudeControl_TS() {}
+
+    // Ensure attitude controller have zero errors to relax rate controller output
+    // Relax only the roll and yaw rate controllers if exclude_pitch is true
+    virtual void relax_attitude_controllers(bool exclude_pitch) override;
+    virtual void input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool plane_controls, float body_roll_cd, float euler_pitch_cd, float euler_yaw_rate_cds) override;
+    void set_is_vectored(bool v) { _is_vectored = v; }
```
Comment (tridge) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:20]: look to be unused?

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduPlane/quadplane.h [hunk 6]
```diff
@@ -115,8 +115,11 @@ class QuadPlane
     // return true when tailsitter frame configured
     bool is_tailsitter(void) const;
 
-    // return true when flying a control surface only tailsitter tailsitter
-    bool is_contol_surface_tailsitter(void) const;
+    // return true when flying a control surface only tailsitter
+    bool is_control_surface_tailsitter(void) const;
+
+    // true when flying a tilt-vectored tailsitter
+    bool _is_vectored;
```
Comment (rmackay9) [inline ArduPlane/quadplane.h:122]: as mentioned on the dev call, there are two variables named, "_is_vectored".  one here in the vechicle code and another in the AC_ATtitude controller.

File: libraries/AC_AttitudeControl/AC_AttitudeControl_TS.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
+#include "AC_AttitudeControl_TS.h"
+
```
Comment (tridge) [inline libraries/AC_AttitudeControl/AC_AttitudeControl_TS.cpp:2]: add a license and comment header

**Comment by rmackay9** [comment]
Just for the record, we decided on the dev call that after the above changes are made this is OK to merge to master.

**Comment by kd0aij** [comment]
resolved issues noted in devcall
1) removed unused member/getter/setter
2) added header to new cpp file
3) ran astyle on new files
and squashed
@Hwurzburg Ready for flight testing.

**Comment by kd0aij** [comment]
@tridge @rmackay9 the 2 new commits change the base class relax_attitude_controllers() method back to non-virtual
and always instantiate AC_AttitudeControl_TS in quadplane, since the new relax_attitude_controllers(bool) method is always used by quadplane.

**Comment by kd0aij** [comment]
@tridge @IamPete1 Henry and I have added another commit to prevent prop strikes when changing to a VTOL mode when armed on the ground. It tests OK in SITL and on one of Henry's TVBS planes.
Not sure why CI is failing now; the failing (copter) tests pass locally, except for replay, and I don't see why replay should be affected by this. Last thing I tried was rebasing on latest master.

**Comment by kd0aij** [comment]
rebased again and passing CI

**Review by tridge — APPROVED** [review]

**Pull request merged**